test/fixtures/dialects/spark3/select_delimited_identifier_with_escaped_backticks.sql

```diff
@@ -0,0 +1 @@
+select `delimited``with escaped` from `some``table`;
```

Comment:
So looking in the [docs](https://spark.apache.org/docs/latest/sql-ref-identifier.html#delimited-identifier):
![image](https://user-images.githubusercontent.com/80432516/145902163-8ff89c13-2c6b-4f23-b637-f63a1facedf0.png)

It looks like in spark that the example above is a single identifier containing backticks in its name: `delimited``with escaped`
But if you refer to how this is being parsed it is being interpreted as a column `delimited` aliased as `with escaped`:
![image](https://user-images.githubusercontent.com/80432516/145902403-733ff0bb-4835-4ce4-b076-599fd168c342.png)

I'm not super familiar with Spark SQL myself so can you clarify a couple things for me:
- Am I correct in thinking that this should be a single `delimited``with escaped` identifier? If so then we'll need a more advanced Regex lexer for back_quote in Spark.
- If yes to the above, what does an aliased back-tick quoted identifier look like in Spark SQL without an AS (if that's possible)? Is this possible 
```
`delimited``with escaped` `alias`
```
I assume it's the fact that those escaped backticks are sequential that differentiates them

Reply:
Ah good point, I didn't notice how it actually parsed. I'll caveat here by saying I am also no expert in this dialect, so I'm just reading the docs and verifying what happens for test queries as I go. 

You are correct that ``` `delimited``with escaped` ``` should be one identifier that contains a backtick; it will output as ``delimited`with escaped``. 

Your example with the alias is also legal. For instance this query will execute under both ANSI mode and not:
```
with a as (
select 1 as `b``c`
)
select `b``c` `!d!`, `b``c` as `?e` from a;
```
The result is: 
![image](https://user-images.githubusercontent.com/9284402/145911773-39680139-22e2-4f06-8713-6f04479acadc.png)

I don't immediately see where I would accommodate this in sqlfluff. 

Can you point me in the right direction or would it be better to keep this PR uncomplicated and raise a separate issue for handling escaped backticks? Per the docs:
> Use ` to escape special characters (e.g., `).

makes me wonder whether there are not other things that should be getting escaped. 

So far I have no examples of a backtick-containing identifier "in the wild", but the need for delimited identifiers has cropped up because of e.g. some naughty systems generating csv's with field names like "ctr, %".

Reply:
Makes sense, you've got to question the design of whoever thought using backticks to escape backticks within backticks was nice design lol 😅 

So in terms of getting this to parse we're going to need to overwrite regex pattern in the back_quote lexer in the ANSI dialect ([here](https://github.com/sqlfluff/sqlfluff/blob/9504fc28af3142f3e4915b0adf5e31746df06e43/src/sqlfluff/dialects/dialect_ansi.py#L85)). The pattern currently matches anything that's not a backtick within 2 backticks.

For Spark we'll need to come up with some regex pattern that matches anything between 2 backticks provided the closing backtick is not immediately preceded or followed by another backtick.

So if you copy the linked ANSI back_quote lexer and add it into Spark [here](https://github.com/sqlfluff/sqlfluff/blob/9504fc28af3142f3e4915b0adf5e31746df06e43/src/sqlfluff/dialects/dialect_spark3.py#L45) then you can start playing around with the regex pattern to match these back_quote containing escaped backticks.

Reply:
In terms of what that pattern would be it's hard to say without playing around myself. 

My initial thoughts would be something like
```
`.*(?<!`)`(?!`)
```
anything between 2 backticks, with negative lookbehind + negative lookahead to check that the closing backtick is not surrounded by another backtick. Can't promise that works but a good place to start.

We use the [regex](https://pypi.org/project/regex/) package so you should have all of the standard python re notations + the additional regex ones at your disposal.

Reply:
(lmk if you're struggling with it though as I'm happy to have a play with the pattern tomorrow. Worst case we can merge as is, given the standard back_quote regex covers 99% of cases, and I'll deal with this edge case later 😄 )

Reply:
Seriously. Ok I'll take a look, thanks!

Reply:
realise my initial pattern suggestion is too greedy. It would capture all of
```
`delimited ``identifier` `alias`
```
rather than just the first possible match.

Making the `.*` non-greedy (i.e. `.*?`) seems to work from my initial testing:
![image](https://user-images.githubusercontent.com/80432516/145986067-bd9016a2-77ba-4cf4-99ed-e01ce4d4f861.png)

so patten would be:
```
`.*?(?<!`)`(?!`)
```

Reply:
heh thanks I am shamefully bad at regex, but that works afaict

Reply:
ah although this also appears to match when it ought to fail i believe:
```
`delimited```with escaped`
```

Reply:
nice, looks good 😄 Could you add some more unit tests to this file to verify that the pattern works for aliases (it does but good to have the tests in case someone breaks it in the future)

```sql
SELECT `delimited ``identifier` `alias` FROM `some``table` `some``alias`;
SELECT `delimited ``identifier` AS `alias` FROM `some``table` AS `some``alias`;
```

Then should be good to go 👍

Reply:
ok, though what do you think of:
```
`.*?(?<![^`]`)`(?!`[^`])
```
to patch the false positive with the triple backticks in my previous comment?

Those are good suggestions I'll add them now.

Reply:
ah nice test case, let me take a look

Reply:
So both the original regex and the patched one you suggested also can't handle a valid column like this
![image](https://user-images.githubusercontent.com/80432516/146011618-75bc0d2f-759c-42ec-9c1a-376407f2eb53.png)

Reply:
Or this too
![image](https://user-images.githubusercontent.com/80432516/146012187-accb9382-5557-4c7e-9c29-80699d715341.png)

Reply:
Okay think I have it: 
```
`([^`]|``)*`
```
opens the backtick quote, greedily matches all the double backtick qoutes and then closes on the next quote. Seems to be able to handle all the test cases, even with multiple escaped backticks, and doesn't parse odd numbers of subsequent ticks 👍 

![image](https://user-images.githubusercontent.com/80432516/146013055-c3fdd63a-85d1-4bf8-96d4-3e6a247a20c4.png)

Might as well chuck these tests in too 😄 
```sql
select ```delimited` `alias` FROM `some``table` `some````alias`;
select `delimited``` `alias` FROM `some``table` ```some````alias```;

SELECT `delimited ``identifier` `alias` FROM `some``table` `some``alias`;
SELECT `delimited ``identifier` AS `alias` FROM `some``table` AS `some``alias`;
```

Reply:
hahaha i could see when i added a third backtick that things could get out of hand
i'll make the amendments and take this out of draft
thanks for your help!

Reply:
haha it's a pandora's box! Thanks for your patience working through this 😄